python/tests/param/params_test.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import, division, print_function
+
+from collections import namedtuple
+# Use this to create parameterized test cases
+from parameterized import parameterized
+
+from sparkdl.param.converters import SparkDLTypeConverters
+
+from ..tests import PythonUnitTestCase
+
+TestCase = namedtuple('TestCase', ['data', 'description'])
+
+_shared_invalid_test_cases = [
+    TestCase(data=['a1', 'b2'], description='required pair but get single element'),
+    TestCase(data=('c3', 'd4'), description='required pair but get single element'),
+    TestCase(data=[('a', 1), ('b', 2)], description='only accept dict, but get list'),
+    TestCase(data={1: 'a', 2.0: 'b'}, description='wrong mapping type'),
+    TestCase(data={'a': 1.0, 'b': 2}, description='wrong mapping type'),
+]
+_col2tnsr_test_cases = _shared_invalid_test_cases + [
+    TestCase(data={'colA': 'tnsrOpA', 'colB': 'tnsrOpB'},
+             description='strict tensor name required'),
+]
+_tnsr2col_test_cases = _shared_invalid_test_cases + [
+    TestCase(data={'tnsrOpA': 'colA', 'tnsrOpB': 'colB'},
+             description='strict tensor name required'),
+]
+
+class ParamsConverterTest(PythonUnitTestCase):
+    """
+    Test MLlib Params introduced in Spark Deep Learning Pipeline
+    Additional test cases are attached via the meta class `TestGenMeta`.
+    """
+
+    def test_tf_input_mapping_converter(self):
+        """ Test valid input mapping conversion """
+        valid_tnsr_input = {'colA': 'tnsrOpA:0', 'colB': 'tnsrOpB:0'}
+        valid_input_mapping_result = [('colA', 'tnsrOpA:0'), ('colB', 'tnsrOpB:0')]
+
+        res = SparkDLTypeConverters.asColumnToTensorNameMap(valid_tnsr_input)
+        self.assertEqual(valid_input_mapping_result, res)
+
+    def test_tf_output_mapping_converter(self):
+        """ Test valid output mapping conversion """
+        valid_tnsr_output = {'tnsrOpA:0': 'colA', 'tnsrOpB:0': 'colB'}
+        valid_output_mapping_result = [('tnsrOpA:0', 'colA'), ('tnsrOpB:0', 'colB')]
+
+        res = SparkDLTypeConverters.asTensorNameToColumnMap(valid_tnsr_output)
+        self.assertEqual(valid_output_mapping_result, res)
+
+    @parameterized.expand(_col2tnsr_test_cases)
```

Comment:
this is much more readable! one question for you -- if the test fails on one of the cases, what would the output look like? will i be able to quickly identify which test case failed?

Reply:
Thanks :D 
The test cases will be appended to the docstring of the base test function.
Here is what it looks like if a test fails. 
```
============= Running the tests in: tests/param/params_test.py =============
Using TensorFlow backend.
Test invalid column name to tensor name mapping [with data=['a1', 'b2'], description='required pair but got single element'] ... ok
Test invalid column name to tensor name mapping [with data=('c3', 'd4'), description='required pair but got single element'] ... ok
Test invalid column name to tensor name mapping [with data=[('a', 1), ('b', 2)], description='only accept dict, but got list'] ... ok
Test invalid column name to tensor name mapping [with data={1: 'a', 2.0: 'b'}, description='wrong mapping type'] ... ok
Test invalid column name to tensor name mapping [with data={'a': 1.0, 'b': 2}, description='wrong mapping type'] ... ok
Test invalid column name to tensor name mapping [with data={'colB': 'tnsrOpB', 'colA': 'tnsrOpA'}, description='tensor name required'] ... ok
Test invalid column name to tensor name mapping [with data={'colB': 'tnsrOpB:1', 'colA': 'tnsrOpA:0'}, description='tensor name required'] ... FAIL
Test invalid tensor name to column name mapping [with data=['a1', 'b2'], description='required pair but got single element'] ... ok
Test invalid tensor name to column name mapping [with data=('c3', 'd4'), description='required pair but got single element'] ... ok
Test invalid tensor name to column name mapping [with data=[('a', 1), ('b', 2)], description='only accept dict, but got list'] ... ok
Test invalid tensor name to column name mapping [with data={1: 'a', 2.0: 'b'}, description='wrong mapping type'] ... ok
Test invalid tensor name to column name mapping [with data={'a': 1.0, 'b': 2}, description='wrong mapping type'] ... ok
Test invalid tensor name to column name mapping [with data={'tnsrOpB': 'colB', 'tnsrOpA': 'colA'}, description='tensor name required'] ... ok
Test valid input mapping conversion ... ok
Test valid output mapping conversion ... ok

======================================================================
FAIL: Test invalid column name to tensor name mapping [with data={'colB': 'tnsrOpB:1', 'colA': 'tnsrOpA:0'}, description='tensor name required']
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/parameterized/parameterized.py", line 392, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/Users/philipyang/CodeBase/spark-deep-learning/python/tests/param/params_test.py", line 71, in test_invalid_input_mapping
    SparkDLTypeConverters.asColumnToTensorNameMap(data)
AssertionError: TypeError not raised

----------------------------------------------------------------------
Ran 15 tests in 0.002s
```